### PR TITLE
From iterator for more tuples

### PR DESCRIPTION
**Pull request by shahn**

### CODE REVIEW

**Comment by shahn** [comment]
this really is just a draft to extend on #132187, which I am not sure if I should push these changes there.

r? libs-api

**Comment by Amanieu** [comment]
@rfcbot fcp merge

This is similar to #132187 and notably allows collecting an iterator of tuples into a separate collection for each tuple element. Support for this already existed for pairs, this PR extends it to arbitrary tuples.

**Comment by shahn** [comment]
With #132187 merged, this should be ready for review now :)

**Comment by shahn** [comment]
@BurntSushi @joshtriplett @m-ou-se friendly ping if you could give your verdict :)

**Review by Amanieu — COMMENTED** [review]
LGTM apart from the indentation issue.

File: library/core/src/iter/traits/collect.rs [hunk 1]
```diff
@@ -595,32 +562,67 @@ macro_rules! spec_tuple_impl {
...
-                    $($var_names: &'a mut impl Extend<$ty_names>,)*
-                ) -> impl FnMut((), ($($ty_names,)*)) + 'a {
-                #[allow(non_snake_case)]
-                // SAFETY: We reserve enough space for the `size_hint`, and the iterator is `TrustedLen`
-                // so its `size_hint` is exact.
-                move |(), ($($extend_ty_names,)*)| unsafe {
-                    $($var_names.extend_one_unchecked($extend_ty_names);)*
+                    fn extend<'a, $($ty_names,)*>(
+                        $($var_names: &'a mut impl Extend<$ty_names>,)*
+                    ) -> impl FnMut((), ($($ty_names,)*)) + 'a {
```
[the remark this answers is not in the document]
Comment (shahn) [reply library/core/src/iter/traits/collect.rs:567]: wow, the formatting is awful in some places, thanks for catching that. I will rebase this with two commits, first one just to fix the formatting, second one with the changes for this patch.

**Comment by shahn** [comment]
Wait what, this auto-closed somehow

**Pull request merged**